### ISSUE TITLE
Suggestions for benchmarking functions

### ISSUE THREAD

**Issue by jrevels** [jrevels/AutoDiffJSoC.jl#1]
@mlubin @scidom I've made this issue to facilitate discussion regarding potential benchmark functions. Feel free to link anybody here who might have a suggestion. 

For those linked here, we're looking for functions of the form `f(x::Vector{T})` that return a scalar `T` (if it returns a scalar of a different type than the input's `eltype`, that could be okay too). Preferably, `f` should work for variable lengths of `x`.

The benchmark we're running is essentially testing how fast we can take the gradient of `f` at a specific input `x`; thus, if `x` is restricted to a certain domain, that would be relevant information so we can properly build the test input vectors.

Thanks for any help you can provide!

**Comment by mlubin** [comment]
@davidanthoff @johnmyleswhite, any suggestions for benchmark functions to test out these improvements to DualNumbers?

**Comment by johnmyleswhite** [comment]
Maybe something like a logistic regression model with N parameters?

**Comment by davidanthoff** [comment]
I don't have a ready function to include. But I did try this in our dynamic programming framework we are working on. Roughly it is characterized by a large number of repeated optimizations (about 30k), using various NLopt solvers. I tried this with a two choice variable problem. Yes, not exactly where this will shine, but that is what I had ready. Here are my perf results:

````
Using Optim.autodiff (i.e. Dual)
   2.657 seconds      (34560 k allocations: 753 MB, 2.46% gc time)

Using NDualTup
   2.281 seconds      (23510 k allocations: 519 MB, 2.06% gc time)

Using NDualVec

  14.095 seconds      (366 M allocations: 15772 MB, 9.60% gc time)
````

All of this with today's 0.4 nightly.

Long story short, the implementation that uses tuples looks promising, the one using vectors not so much.

**Comment by jrevels** [comment]
> I tried this with a two choice variable problem.

@davidanthoff Does this mean that your input vector size was length 2, or am I misunderstanding that? What was the `N` parameter of the types for those benchmarks you ran (i.e. number of epsilon components)? If you are able to share, I'd be really interested in reading the code you benchmarked with.

Generally speaking, our initial findings indicate that  `NDualTup` could be the better option for low input vector sizes, while `NDualVec` scales much better to higher sizes, as it is not limited by stack memory and - I believe - doesn't lean as hard on the cache.

**Comment by davidanthoff** [comment]
@jrevels Yes, ``N=2`` in the these runs. I think I can fairly easily modify the problem to run with ``N=4``, maybe even ``N=5`` and report back those results as well.

Our algorithm is characterized by low input vector sizes, but a very large number of computations. The 30k calls to ``solve`` lead to about 280k evaluations of the objective function and about 270k evaluations of the gradient of the objective function. Our objective function is also fairly computationally intensive: it includes the evaluation of a function approximation that is a tensor-basis with chebychev polynomials and that ends up being a lot of operations.

The code is very much work in progress right now and not public. I'd be happy to give you read access, but please don't share it or use it for research/publications or anything really at this point. We intend to open source it eventually. If you send me your bitucket account name I'll give you access.

**Comment by jrevels** [comment]
> @jrevels Yes, N=2 in the these runs. I think I can fairly easily modify the problem to run with N=4, maybe even N=5 and report back those results as well.

If you have the time, I'm always interested in seeing more benchmarks! 

> Our algorithm is characterized by low input vector sizes, but a very large number of computations.

I suspect your results with `NDualTup` and `NDualVec` might end up being pretty typical for this use case, then. 

> The code is very much work in progress right now and not public. I'd be happy to give you read access, but please don't share it or use it for research/publications or anything really at this point. We intend to open source it eventually. If you send me your bitucket account name I'll give you access.

Understood. Here's what I would want to do should you decide to share it with me:

1. Read it so I can get a feel for how you're using the `NDual*` types and the `autodiff` function, and what the bottlenecks end up being our side.
2. Perhaps use it as a test case for benchmarks moving forward (privately, of course; it wouldn't go on GitHub or anything)
3. Discuss the code/benchmark results with my mentors for this project, @mlubin and @scidom. 

If you're uncomfortable sharing the code for any of the above purposes, I understand completely, and I'd be happy to at least see more benchmark results even if I can't look at the code that generated them. My Bitbucket username is the same as my GitHub username (just made an account). Also, feel free to email me at the address linked to my GitHub account should you have any questions/concerns/etc.

**Comment by davidanthoff** [comment]
Just gave you access. Everything you suggest is perfectly fine, we just don't want anyone to come out and publish a paper with this before we do ;)

**Comment by mlubin** [comment]
14% speedup by going from ``N=1`` (Dual) to ``N=2`` isn't too shabby. The best you could possibly hope for is 50%

**Comment by davidanthoff** [comment]
No, not shabby at all, actually really good! The tuple implementation doesn't utilize SIMD right now, correct?

**Comment by mlubin** [comment]
Correct, afaik. Very interested to see how that changes things, but seems to be beyond our control at this point.

**Comment by davidanthoff** [comment]
Are the people working on tuples/SIMD aware of this use-case, the barriers to getting this to work? This use-case seems to be one of the nicest ones for SIMD, I would expect that essentially every single  operation could be turned into SIMD instructions in the case of forward diff.

**Comment by mlubin** [comment]
Feel free to raise it here: https://github.com/JuliaLang/julia/issues/11899. It's blocked by a major codegen rewrite which seemingly won't get into 0.4.

**Comment by davidanthoff** [comment]
Ah, didn't see that. Well, I don't want to suggest anything that might hold 0.4 up any longer, so I'll just keep quiet for now ;)
